### PR TITLE
Handle race conditions involving rename in LocalFileSystem.

### PR DESCRIPTION
**Pull request by htran1**
One race condition is when a directory D1/a/b/c/d is renamed to D2/a/b/c/d after another thread created D2/a/b/c/d.
This resulted in D2/a/b/c/d/d. Handle this by using the Java renameTo() method to avoid the LocalFileSystem fallback of copy.
The copy fallback also results in a race condition when a file is opened for copy in one thread and the parent directory is createed in another thread. A file not found exception is encountered in this case. This case is handled by catching the exception and retrying a few times.

### CODE REVIEW

**Comment by htran1** [comment]
@chavdar can you review?

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java [hunk 1]
```diff
@@ -549,6 +552,26 @@ public static boolean unsafeRenameIfNotExists(FileSystem fs, Path from, Path to)
       if (!fs.exists(to.getParent())) {
         fs.mkdirs(to.getParent());
       }
+
+      boolean renamed;
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: This does not seem to be used. Please remove.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java [hunk 2]
```diff
@@ -549,6 +552,26 @@ public static boolean unsafeRenameIfNotExists(FileSystem fs, Path from, Path to)
...
+        if (fs instanceof LocalFileSystem) {
+          File srcFile = ((LocalFileSystem) fs).pathToFile(from);
+          File dstFile = ((LocalFileSystem) fs).pathToFile(to);
+
+          // failure to rename directory occurs when target name already exists
+          return srcFile.renameTo(dstFile);
+        }
+        else if (fs instanceof RateControlledFileSystem &&
+              ((RateControlledFileSystem) fs).getDecoratedObject() instanceof LocalFileSystem) {
+          return fs.rename(from, to);
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: Why not use the above method?

File: gobblin-utility/src/main/java/gobblin/util/RateControlledFileSystem.java [hunk 3]
```diff
@@ -139,8 +141,24 @@ public boolean mkdirs(Path path, FsPermission permission) throws IOException {
 
   @Override
   public boolean rename(Path path0, Path path1) throws IOException {
+    boolean renamed;
+
     this.acquirePermit();
-    return this.fs.rename(path0, path1);
+
+    // Use Java rename when renaming a directory for LocalFileSystem.
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/RateControlledFileSystem.java]: This code appears in two places. Please consider refactoring it into a function.

**Pull request merged**